Add testmempoolaccept endpoint

mononaut opened the pull request:
This PR adds a new REST API endpoint `POST /txs/test`, which accepts a comma-separated list of transactions as raw hex strings, passes them to `testmempoolaccept`, and returns the results as JSON.

mononaut committed:
Add testmempoolaccept endpoint

junderw commented on src/daemon.rs:
```diff
@@ -117,6 +117,26 @@ struct NetworkInfo {
+pub struct MempoolAcceptResult {
+    txid: String,
+    wtxid: String,
+    allowed: Option<bool>,
```
It might be a good idea to document what `None` means for allowed.

This will be serialized as no allowed key when sent to the client... which is weird.

junderw replied:
Looks like it will be `None` when the transaction was not even validated. (ie. because a parent in the same request already returned false for allowed....

We could collapse all None into false imo...... but I mean, however you want to do it is fine.

junderw replied:
Apparently you can verify unconfirmed packages but you need to place them in the array in dependency order, so the parent always comes before the child.

It's also kind of weird that allowed might be false because of having too high of a fee.

Depending on what we want to do here, it might be a good idea to add the maxfeerate option as well. (default 0.1 BTC/kvB)

mononaut replied:
> This will be serialized as no allowed key when sent to the client... which is weird.

you would think so, but it actually seems to get serialized as `{ ..., allowed: null, ... }`.

mononaut replied:
Collapsing to false does seem sensible, but if the key being omitted conveys some information that isn't apparent otherwise (i.e. that the transaction is disallowed because an ancestor failed validation), then I'm reluctant to remove that.

mononaut replied:
> Depending on what we want to do here, it might be a good idea to add the maxfeerate option as well. (default 0.1 BTC/kvB)

yes, absolutely

junderw commented:
Q: Where do you plan to use this endpoint and how?

mononaut commented:
Access to `testmempoolaccept` is a frequent feature request from some of our API users, but I'll probably also add an interface to the mempool frontend similar to the `/tx/push` page.

junderw commented:
Yeah, if we're going to offer this publicly, we should definitely offer an optional maxfeerate parameter.

Might be worth it to check the request body to make sure it's only hex and commas... and maybe we should limit it somehow. (Might be a DoS vector)

mononaut committed:
testmempoolaccept add maxfeerate param

mononaut committed:
Add testmempoolaccept pre-checks

mononaut commented:
Added support for the optional maxfeerate parameter (via query string), and added some basic pre-checks to the POSTed raw transactions for valid hexadecimal-ness and string length.

junderw commented on src/daemon.rs:
```diff
@@ -582,6 +602,20 @@ impl Daemon {
+        maxfeerate: Option<f32>,
+    ) -> Result<Vec<MempoolAcceptResult>> {
+        let params = match maxfeerate {
+            Some(rate) => json!([txhex, format!("{:.8}", rate)]),
```
f32::EPSILON (the smallest difference representable above 1.0) works out to about 11 sats when we consider 1.0 to be 1 BTC. So it can only really reliably show 7 decimal places.

Which I guess is fine if we consider 0.00001 BTC /kvB (5 decimal places) to be the minimum. (1 sat per vB) In which case showing 8 decimal places is pointless.

I wonder what the type is in Core. Might be worth it to use that, and round it to 8 if we use f64 and 7 if we use f32.

junderw replied:
https://doc.rust-lang.org/std/primitive.f32.html#associatedconstant.EPSILON

`0.00000011920929` is the smallest increment of f32.

junderw replied:
https://doc.rust-lang.org/std/primitive.f64.html#associatedconstant.EPSILON

`0.00000000000000022204460492503131` is smallest for f64

junderw replied:
Or we could always just accept the sane sat/vB instead of BTC/kvB, and just slide the decimal place 5 to the left. lol

mononaut replied:
testmempoolaccept in Core throws an error if you give it too many decimal places (e.g. from floating point rounding errors), so it needs to be truncated to something.

I picked 8.d.p out of habit, but yeah I'll check what Core's bounds actually are.

mononaut replied:
it turns out 8.d.p is actually exactly Core's limit (it's parsed as a fixed point number with scale 8):
https://github.com/bitcoin/bitcoin/blob/c1223188e0a5fb11c3a1b9224511a49dc2f848ed/src/rpc/util.cpp#L82-L87
https://github.com/bitcoin/bitcoin/blob/c1223188e0a5fb11c3a1b9224511a49dc2f848ed/src/rpc/util.h#L105

junderw replied:
In that case you should use f64 so that 8 decimal points can be properly represented without truncating.

junderw commented on src/rest.rs:
```diff
@@ -1202,6 +1202,38 @@ fn handle_request(
+            txhexes.iter().try_for_each(|txhex| {
+                // each transaction must be of reasonable size (more than 60 bytes, within 400kWU standardness limit)
+                if !(120..800_000).contains(&txhex.len()) {
+                    Result::Err(HttpError::from("Invalid transaction size".to_string()))
```
Maybe it might be kind to return the 0-base index of the transaction that caused the error.

junderw commented on src/rest.rs:
```diff
@@ -1202,6 +1202,38 @@ fn handle_request(
             http_message(StatusCode::OK, txid.to_hex(), 0)
         }
+        (&Method::POST, Some(&"txs"), Some(&"test"), None, None, None) => {
+            let txhexes: Vec<String> = String::from_utf8(body.to_vec())?
```
It might be a good idea to limit the number of txes per request.

junderw commented on src/rest.rs:
```diff
@@ -1202,6 +1202,38 @@ fn handle_request(
         }
+        (&Method::POST, Some(&"txs"), Some(&"test"), None, None, None) => {
+            let txhexes: Vec<String> = String::from_utf8(body.to_vec())?
+                .split(',')
```
From an API user perspective, does raw text hex separated by commas make more sense than a JSON array of strings?

Obviously this adds some unnecessary `[]` and `""` surrounding everything, but it seems a bit weird to just have raw stuff with random commas.

I guess this was just an extension of the existing push tx endpoint, and then commas were tacked on.

Maybe `\n` might be a better separator? Or `:` or something that doesn't make people wonder why it wasn't just a JSON to begin with.

idk, this isn't really a big issue, I think it's fine as is... but while the push tx endpoint feels fine, it just feels kind of weird when you add in multiple txes with commas.

mononaut replied:
hmm, I was thinking that comma-separated strings were more framework-agnostic. But the response is JSON, so I guess we can assume users have that capability 😅

mononaut committed:
testmempoolaccept JSON input, tx limit, error indices

mononaut commented:
- Switched the input format to a JSON serialized list of raw hex strings
- Enforced a maximum of 25 transactions per request (to match Core's limit)
- Include the index of the transaction in the error response when per-tx precheck validation fails.

mononaut committed:
testmempoolaccept maxfeerate f32 -> f64

mononaut commented:
Switched maxfeerate param from f32 to f64

softsimon reviewed (approved):
tested ACK @ [ac32e4b1c3]

` % curl -X POST -H "Content-Type: application/json" -d '["02000000000101a89d6d6e5475d9f17f153bd71c983170e6eb7d782966021ffea8e8bf03b791d50000000000fdffffff010d2a00000000000016001449b41a4c9fdd27da6d2c182078cbcd948b8e275a0247304402200d38e7b028f0f92778f3a7eca7208eee8a8fa397e7a1970eda63c5b6c0103e900220141239cc25d7b6e7ae6df833037d8d40b869d1d9ea57c0f71e516349eb9e0099012103de74b84d2a5715f317c23b84520a76a2b7ea5f7ba6302bd64ddf439eba546e11dd762700"]' http://localhost:3000/txs/test`
 
`[{"txid":"c27cdcc1b07096e8bc0b29c9868a4052231548c0b23f20d5437226baa42e004e","wtxid":"16586c10f96e52410c9a62ae80e5f8891b62ca887ff9722fa45653e355bdfc60","allowed":true,"vsize":110,"fees":{"base":0.00001562,"effective-feerate":0.000142,"effective-includes":["16586c10f96e52410c9a62ae80e5f8891b62ca887ff9722fa45653e355bdfc60"]},"reject-reason":null}]%`
